### ISSUE TITLE
[Bug] HeteroData.to_homogeneous() doesn't handle the train/val/test mask correctly

### ISSUE THREAD

**Issue by junhongmit** [pyg-team/pytorch_geometric#8856]
### 🐛 Describe the bug

When converting a heterogeneous graph to a homogeneous graph using HeteroData.to_homogeneous(), the train/val/test mask is not masking the correct node. The following is an example to reproduce it:
```Python
from torch_geometric.datasets import OGB_MAG
dataset = OGB_MAG()
data = dataset[0]
homo = data.to_homogeneous()
print(data['paper'].y[data['paper'].train_mask]) # This is the correct masked training set
print(homo.y[homo.train_mask]) # You will see -1 in the output label, which means irrelevant node got masked. OGB_MAG has -1 label for those nodes not in the train/val/test set.
print(data['paper'].y[data['paper'].train_mask].sum() == homo.y[homo.train_mask].sum()) # And the label sum does not equal
```
The above code snippet generates the following output:
```
tensor([246, 131, 189,  ..., 266, 289,   1]) # This is the correct masked training set
tensor([246, 131, 189,  ...,  -1,  -1,  -1]) # You will see -1 in the output label, which means irrelevant node got masked. OGB_MAG has -1 label for those nodes not in the train/val/test set.
tensor(False) # And the label sum does not equal
```

### Versions

PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.4
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.3 (Ootpa) (ppc64le)
GCC version: (GCC) 8.3.1 20191121 (Red Hat 8.3.1-5)
Clang version: Could not collect
CMake version: version 3.11.4
Libc version: glibc-2.28

Python version: 3.9.18 (main, Sep 11 2023, 14:34:07)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-240.el8.ppc64le-ppc64le-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.4.152
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 535.86.10
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:         ppc64le
Byte Order:           Little Endian
CPU(s):               160
On-line CPU(s) list:  0,1,4,5,8,9,12,13,16,17,20,21,24,25,28,29,32,33,36,37,40,41,44,45,48,49,52,53,56,57,60,61,64,65,68,69,72,73,76,77,80,81,84,85,88,89,92,93,96,97,100,101,104,105,108,109,112,113,116,117,120,121,124,125,128,129,132,133,136,137,140,141,144,145,148,149,152,153,156,157
Off-line CPU(s) list: 2,3,6,7,10,11,14,15,18,19,22,23,26,27,30,31,34,35,38,39,42,43,46,47,50,51,54,55,58,59,62,63,66,67,70,71,74,75,78,79,82,83,86,87,90,91,94,95,98,99,102,103,106,107,110,111,114,115,118,119,122,123,126,127,130,131,134,135,138,139,142,143,146,147,150,151,154,155,158,159
Thread(s) per core:   2
Core(s) per socket:   20
Socket(s):            2
NUMA node(s):         6
Model:                2.1 (pvr 004e 1201)
Model name:           POWER9, altivec supported
CPU max MHz:          3800.0000
CPU min MHz:          2300.0000
L1d cache:            32K
L1i cache:            32K
L2 cache:             512K
L3 cache:             10240K
NUMA node0 CPU(s):    0,1,4,5,8,9,12,13,16,17,20,21,24,25,28,29,32,33,36,37,40,41,44,45,48,49,52,53,56,57,60,61,64,65,68,69,72,73,76,77
NUMA node8 CPU(s):    80,81,84,85,88,89,92,93,96,97,100,101,104,105,108,109,112,113,116,117,120,121,124,125,128,129,132,133,136,137,140,141,144,145,148,149,152,153,156,157
NUMA node252 CPU(s):  
NUMA node253 CPU(s):  
NUMA node254 CPU(s):  
NUMA node255 CPU(s):

Versions of relevant libraries:
[pip3] botorch==0.8.5
[pip3] gpytorch==1.10
[pip3] numpy==1.23.5
[pip3] numpydoc==1.6.0
[pip3] performer-pytorch==1.1.4
[pip3] pytorch-lightning==2.1.3
[pip3] torch==1.12.1
[pip3] torch-cluster==1.6.3
[pip3] torch-geometric==2.3.0
[pip3] torch-scatter==2.1.2
[pip3] torch-sparse==0.6.18
[pip3] torchmetrics==0.11.4
[pip3] torchtext==0.13.1a0+35066f2
[pip3] torchvision==0.13.1
[conda] _pytorch_select           2.0                      cuda_2    https://ftp.osuosl.org/pub/open-ce/current
[conda] botorch                   0.8.5                         0    pytorch
[conda] cudatoolkit               11.4.4              h5d40d8d_10    https://opence.mit.edu
[conda] cudatoolkit-dev           11.4.4               h13ae079_2    https://opence.mit.edu
[conda] gpytorch                  1.10                          0    gpytorch
[conda] linear_operator           0.4.0                         0    gpytorch
[conda] numpy                     1.23.5           py39h181cc9a_0  
[conda] numpy-base                1.23.5           py39h1bde650_0  
[conda] numpydoc                  1.6.0                    pypi_0    pypi
[conda] performer-pytorch         1.1.4                    pypi_0    pypi
[conda] pytorch                   1.12.1          cuda11.4_py39_1    https://opence.mit.edu
[conda] pytorch-base              1.12.1          cuda11.4_py39_pb3.19_2    https://opence.mit.edu
[conda] pytorch-lightning         2.1.3                    pypi_0    pypi
[conda] pytorch_geometric         2.4.0              pyhd8ed1ab_0    conda-forge
[conda] torch-cluster             1.6.3                    pypi_0    pypi
[conda] torch-geometric           2.3.0                    pypi_0    pypi
[conda] torch-scatter             2.1.2                    pypi_0    pypi
[conda] torch-sparse              0.6.18                   pypi_0    pypi
[conda] torchmetrics              1.1.2                    pypi_0    pypi
[conda] torchtext-base            0.13.1          cuda11.4_py39_1    https://opence.mit.edu
[conda] torchvision-base          0.13.1          cuda11.4_py39_1    https://opence.mit.edu

**Comment by rusty1s** [comment]
Nice catch :) Will be fixed in https://github.com/pyg-team/pytorch_geometric/pull/8858
